File: import_dump.py
#!/usr/bin/env python3
"""Script to import JSON dump data into MongoDB for Greeters CMS"""

import json
import asyncio
from motor.motor_asyncio import AsyncIOMotorClient
from datetime import datetime
import os
from pathlib import Path

# Load environment variables
from dotenv import load_dotenv
load_dotenv(Path(__file__).parent / 'backend' / '.env')

MONGO_URL = os.environ.get('MONGO_URL', 'mongodb://localhost:27017')
DB_NAME = os.environ.get('DB_NAME', 'greeters_cms')

async def import_data(dump_file: str):
    """Import data from JSON dump file to MongoDB"""
    
    # Connect to MongoDB
    client = AsyncIOMotorClient(MONGO_URL)
    db = client[DB_NAME]
    
    # Read the dump file
    with open(dump_file, 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    print(f"Connected to MongoDB: {DB_NAME}")
    
    # Import users
    if 'users' in data:
        users = data['users']['rows']
        if users:
            # Drop existing collection
            await db.users.drop()
            # Transform and insert
            for user in users:
                user_doc = {
                    'id': user['id'],
                    'email': user['email'],
                    'name': user['name'],
                    'password_hash': user['password_hash'],
                    'role': user['role'],
                    'created_by': user.get('created_by'),
                    'created_at': user['created_at']
                }
                await db.users.insert_one(user_doc)
            print(f"Imported {len(users)} users")
    
    # Import sessions
    if 'sessions' in data:
        sessions = data['sessions']['rows']
        if sessions:
            await db.sessions.drop()
            for session in sessions:
                session_doc = {
                    'id': session['id'],
                    'user_id': session['user_id'],
                    'token_hash': session['token_hash'],
                    'expires_at': session['expires_at'],
                    'created_at': session['created_at']
                }
                await db.sessions.insert_one(session_doc)
            print(f"Imported {len(sessions)} sessions")
    
    # Import pages
    if 'pages' in data:
        pages = data['pages']['rows']
        if pages:
            await db.pages.drop()
            for page in pages:
                page_doc = {
                    'id': page['id'],
                    'locale': page.get('locale', 'fr'),
                    'title': page['title'],
                    'slug': page['slug'],
                    'metaDescription': page.get('meta_description'),
                    'metaKeywords': page.get('meta_keywords'),
                    'isInMenu': page.get('is_in_menu', False),
                    'menuOrder': page.get('menu_order', 0),
                    'menuLabel': page.get('menu_label'),
                    'sections': page.get('sections', []),
                    'status': page.get('status', 'published'),
                    'created_by': page.get('created_by'),
                    'created_at': page.get('created_at'),
                    'updated_at': page.get('updated_at')
                }
                await db.pages.insert_one(page_doc)
            print(f"Imported {len(pages)} pages")
    
    # Import AI chat sessions
    if 'ai_chat_sessions' in data:
        ai_sessions = data['ai_chat_sessions']['rows']
        if ai_sessions:
            await db.ai_chat_sessions.drop()
            for session in ai_sessions:
                session_doc = {
                    'id': session['id'],
                    'created_by': session['created_by'],
                    'locale': session.get('locale', 'fr'),
                    'title': session.get('title'),
                    'latest_draft': session.get('latest_draft'),
                    'created_at': session['created_at'],
                    'updated_at': session.get('updated_at', session['created_at'])
                }
                await db.ai_chat_sessions.insert_one(session_doc)
            print(f"Imported {len(ai_sessions)} AI chat sessions")
    
    # Import AI chat messages
    if 'ai_chat_messages' in data:
        ai_messages = data['ai_chat_messages']['rows']
        if ai_messages:
            await db.ai_chat_messages.drop()
            for msg in ai_messages:
                msg_doc = {
                    'id': msg['id'],
                    'session_id': msg['session_id'],
                    'role': msg['role'],
                    'content': msg['content'],
                    'generated_page': msg.get('generated_page'),
                    'created_at': msg['created_at']
                }
                await db.ai_chat_messages.insert_one(msg_doc)
            print(f"Imported {len(ai_messages)} AI chat messages")
    
    # Create indexes
    await db.users.create_index('email', unique=True)
    await db.users.create_index('id', unique=True)
    await db.pages.create_index([('locale', 1), ('slug', 1)])
    await db.pages.create_index([('locale', 1), ('status', 1)])
    await db.pages.create_index('id', unique=True)
    await db.sessions.create_index('user_id')
    await db.sessions.create_index('id', unique=True)
    await db.ai_chat_sessions.create_index('id', unique=True)
    await db.ai_chat_messages.create_index('session_id')
    await db.ai_chat_messages.create_index('id', unique=True)
    
    print("Created indexes")
    
    # Print summary
    print("\n=== Import Summary ===")
    print(f"Users: {await db.users.count_documents({})}")
    print(f"Sessions: {await db.sessions.count_documents({})}")
    print(f"Pages: {await db.pages.count_documents({})}")
    print(f"AI Sessions: {await db.ai_chat_sessions.count_documents({})}")
    print(f"AI Messages: {await db.ai_chat_messages.count_documents({})}")
    
    client.close()
    print("\nImport completed successfully!")

if __name__ == "__main__":
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: python import_dump.py <dump_file.json>")
        sys.exit(1)
    
    dump_file = sys.argv[1]
    if not os.path.exists(dump_file):
        print(f"Error: File not found: {dump_file}")
        sys.exit(1)
    
    asyncio.run(import_data(dump_file))

File: greeters/scripts/import-dump.ts
import "dotenv/config";

import fs from "node:fs/promises";
import path from "node:path";

import {
  AiChatRole,
  PageStatus,
  PreviewStatus,
  UserRole,
  type Prisma,
} from "@prisma/client";

import { prisma } from "../lib/db/prisma";

type DumpRow = Record<string, unknown>;
type DumpTable = { rows?: DumpRow[] };
type DumpPayload = Record<string, DumpTable>;

const DUMP_PATH = process.argv[2] ?? path.join(process.cwd(), "dump.json");

const readDump = async () => JSON.parse(await fs.readFile(DUMP_PATH, "utf-8")) as DumpPayload;

const asString = (value: unknown) => (typeof value === "string" ? value : "");
const asOptionalString = (value: unknown) => (typeof value === "string" && value.length > 0 ? value : null);
const asNumber = (value: unknown, fallback = 0) => (typeof value === "number" ? value : fallback);
const asOptionalNumber = (value: unknown) => (typeof value === "number" ? value : null);
const asBoolean = (value: unknown, fallback = false) => (typeof value === "boolean" ? value : fallback);
const asDate = (value: unknown) => (typeof value === "string" ? new Date(value) : new Date());
const asOptionalDate = (value: unknown) => (typeof value === "string" ? new Date(value) : null);
const asJson = (value: unknown) => (value ?? null) as Prisma.InputJsonValue;

const toUserRole = (value: unknown) => {
  switch (asString(value).toLowerCase()) {
    case "super_admin":
      return UserRole.SUPER_ADMIN;
    case "admin":
      return UserRole.ADMIN;
    default:
      return UserRole.EDITOR;
  }
};

const toPageStatus = (value: unknown) => {
  switch (asString(value).toLowerCase()) {
    case "published":
      return PageStatus.PUBLISHED;
    case "pending":
      return PageStatus.PENDING;
    case "archived":
      return PageStatus.ARCHIVED;
    default:
      return PageStatus.DRAFT;
  }
};

const toPreviewStatus = (value: unknown) => {
  switch (asString(value).toLowerCase()) {
    case "validated":
      return PreviewStatus.VALIDATED;
    case "rejected":
      return PreviewStatus.REJECTED;
    case "expired":
      return PreviewStatus.EXPIRED;
    default:
      return PreviewStatus.PENDING;
  }
};

const toAiChatRole = (value: unknown) => (asString(value).toLowerCase() === "assistant" ? AiChatRole.ASSISTANT : AiChatRole.USER);

const chunk = <T>(rows: T[], size = 100) => {
  const chunks: T[][] = [];

  for (let index = 0; index < rows.length; index += size) {
    chunks.push(rows.slice(index, index + size));
  }

  return chunks;
};

async function createManyInChunks<T>(rows: T[], handler: (data: T[]) => Promise<unknown>) {
  for (const batch of chunk(rows)) {
    if (batch.length > 0) {
      await handler(batch);
    }
  }
}

async function main() {
  const dump = await readDump();

  console.log(`Import du dump depuis ${DUMP_PATH}`);
  console.log("Réinitialisation des tables cibles...");

  await prisma.aiChatMessage.deleteMany();
  await prisma.aiChatSession.deleteMany();
  await prisma.pageEdit.deleteMany();
  await prisma.pagePreview.deleteMany();
  await prisma.pageVersion.deleteMany();
  await prisma.pageContent.deleteMany();
  await prisma.session.deleteMany();
  await prisma.passwordReset.deleteMany();
  await prisma.homeSection.deleteMany();
  await prisma.menu.deleteMany();
  await prisma.document.deleteMany();
  await prisma.page.deleteMany();
  await prisma.user.deleteMany();

  const users = (dump.users?.rows ?? []).map((row) => ({
    id: asString(row.id),
    email: asString(row.email),
    name: asString(row.name),
    passwordHash: asString(row.password_hash),
    role: toUserRole(row.role),
    createdBy: asOptionalString(row.created_by),
    createdAt: asDate(row.created_at),
  }));
  await createManyInChunks(users, (data) => prisma.user.createMany({ data }));
  console.log(`Users importés: ${users.length}`);

  const sessions = (dump.sessions?.rows ?? []).map((row) => ({
    id: asString(row.id),
    userId: asString(row.user_id),
    tokenHash: asString(row.token_hash),
    expiresAt: asDate(row.expires_at),
    createdAt: asDate(row.created_at),
  }));
  await createManyInChunks(sessions, (data) => prisma.session.createMany({ data }));
  console.log(`Sessions importées: ${sessions.length}`);

  const passwordResets = (dump.password_resets?.rows ?? []).map((row) => ({
    id: asString(row.id),
    userId: asString(row.user_id),
    email: asString(row.email),
    tokenHash: asString(row.token_hash),
    expiresAt: asDate(row.expires_at),
    createdAt: asDate(row.created_at),
  }));
  await createManyInChunks(passwordResets, (data) => prisma.passwordReset.createMany({ data }));
  console.log(`Password resets importés: ${passwordResets.length}`);

  const pages = (dump.pages?.rows ?? []).map((row) => ({
    id: asString(row.id),
    locale: asString(row.locale) || "fr",
    title: asString(row.title),
    slug: asString(row.slug),
    metaTitle: asOptionalString(row.meta_title),
    metaDescription: asOptionalString(row.meta_description),
    metaKeywords: asOptionalString(row.meta_keywords),
    canonicalUrl: asOptionalString(row.canonical_url),
    robotsDirective: asOptionalString(row.robots_directive),
    ogTitle: asOptionalString(row.og_title),
    ogDescription: asOptionalString(row.og_description),
    ogImageUrl: asOptionalString(row.og_image_url),
    ogImageAlt: asOptionalString(row.og_image_alt),
    twitterTitle: asOptionalString(row.twitter_title),
    twitterDescription: asOptionalString(row.twitter_description),
    twitterImageUrl: asOptionalString(row.twitter_image_url),
    focusKeyword: asOptionalString(row.focus_keyword),
    secondaryKeywords: asOptionalString(row.secondary_keywords),
    schemaOrgJson: asOptionalString(row.schema_org_json),
    imageRecommendations: asJson(row.image_recommendations ?? []),
    sitemapPriority: asOptionalNumber(row.sitemap_priority),
    sitemapChangeFreq: asOptionalString(row.sitemap_change_freq),
    sections: asJson(row.sections ?? []),
    status: toPageStatus(row.status),
    isInMenu: asBoolean(row.is_in_menu),
    menuOrder: asNumber(row.menu_order),
    menuLabel: asOptionalString(row.menu_label),
    currentVersion: asNumber(row.current_version, 1),
    publishedVersion: asOptionalNumber(row.published_version),
    createdBy: asString(row.created_by),
    createdAt: asDate(row.created_at),
    updatedBy: asOptionalString(row.updated_by),
    updatedAt: asOptionalDate(row.updated_at),
  }));
  await createManyInChunks(pages, (data) => prisma.page.createMany({ data }));
  console.log(`Pages importées: ${pages.length}`);

  const pageVersions = (dump.page_versions?.rows ?? []).map((row) => ({
    id: asString(row.id),
    pageId: asString(row.page_id),
    versionNumber: asNumber(row.version_number, 1),
    content: asJson(row.content ?? {}),
    status: toPageStatus(row.status),
    createdBy: asString(row.created_by),
    createdAt: asDate(row.created_at),
    approvedBy: asOptionalString(row.approved_by),
    approvedAt: asOptionalDate(row.approved_at),
    rejectionReason: asOptionalString(row.rejection_reason),
  }));
  await createManyInChunks(pageVersions, (data) => prisma.pageVersion.createMany({ data }));
  console.log(`Versions importées: ${pageVersions.length}`);

  const menus = (dump.menus?.rows ?? []).map((row) => ({
    id: asString(row.id),
    items: asJson(row.items ?? []),
    updatedBy: asOptionalString(row.updated_by),
    updatedAt: asOptionalDate(row.updated_at),
  }));
  await createManyInChunks(menus, (data) => prisma.menu.createMany({ data }));
  console.log(`Menus importés: ${menus.length}`);

  const documents = (dump.documents?.rows ?? []).map((row) => ({
    id: asString(row.id),
    filename: asString(row.filename),
    originalFilename: asString(row.original_filename),
    filePath: asString(row.file_path),
    fileSize: BigInt(asNumber(row.file_size)),
    mimeType: asString(row.mime_type),
    category: asString(row.category),
    description: asOptionalString(row.description),
    uploadedBy: asString(row.uploaded_by),
    createdAt: asDate(row.created_at),
  }));
  await createManyInChunks(documents, (data) => prisma.document.createMany({ data }));
  console.log(`Documents importés: ${documents.length}`);

  const homeSections = (dump.home_sections?.rows ?? []).map((row) => ({
    id: asString(row.id),
    sectionType: asString(row.section_type),
    content: asJson(row.content),
    items: asJson(row.items),
    order: asNumber(row.order),
    updatedAt: asDate(row.updated_at),
  }));
  await createManyInChunks(homeSections, (data) => prisma.homeSection.createMany({ data }));
  console.log(`Sections home importées: ${homeSections.length}`);

  const pageContents = (dump.page_contents?.rows ?? []).map((row) => ({
    id: asString(row.id),
    pageId: asString(row.page_id),
    content: asJson(row.content ?? {}),
    updatedAt: asDate(row.updated_at),
  }));
  await createManyInChunks(pageContents, (data) => prisma.pageContent.createMany({ data }));
  console.log(`Contenus de page importés: ${pageContents.length}`);

  const pagePreviews = (dump.page_previews?.rows ?? []).map((row) => ({
    id: asString(row.id),
    pageId: asString(row.page_id),
    newContent: asJson(row.new_content ?? {}),
    status: toPreviewStatus(row.status),
    createdBy: asString(row.created_by),
    createdAt: asDate(row.created_at),
  }));
  await createManyInChunks(pagePreviews, (data) => prisma.pagePreview.createMany({ data }));
  console.log(`Previews importées: ${pagePreviews.length}`);

  const pageEdits = (dump.page_edits?.rows ?? []).map((row) => ({
    id: asString(row.id),
    pageId: asString(row.page_id),
    prompt: asString(row.prompt),
    changesSummary: asOptionalString(row.changes_summary),
    editorId: asOptionalString(row.editor_id),
    editorName: asOptionalString(row.editor_name),
    createdAt: asDate(row.created_at),
  }));
  await createManyInChunks(pageEdits, (data) => prisma.pageEdit.createMany({ data }));
  console.log(`Éditions importées: ${pageEdits.length}`);

  const aiChatSessions = (dump.ai_chat_sessions?.rows ?? []).map((row) => ({
    id: asString(row.id),
    createdBy: asString(row.created_by),
    locale: asString(row.locale) || "fr",
    title: asOptionalString(row.title),
    latestDraft: asJson(row.latest_draft),
    createdAt: asDate(row.created_at),
    updatedAt: asDate(row.updated_at),
  }));
  await createManyInChunks(aiChatSessions, (data) => prisma.aiChatSession.createMany({ data }));
  console.log(`Sessions IA importées: ${aiChatSessions.length}`);

  const aiChatMessages = (dump.ai_chat_messages?.rows ?? []).map((row) => ({
    id: asString(row.id),
    sessionId: asString(row.session_id),
    role: toAiChatRole(row.role),
    content: asString(row.content),
    generatedPage: asJson(row.generated_page),
    createdAt: asDate(row.created_at),
  }));
  await createManyInChunks(aiChatMessages, (data) => prisma.aiChatMessage.createMany({ data }));
  console.log(`Messages IA importés: ${aiChatMessages.length}`);

  console.log("Import terminé avec succès.");
}

main()
  .catch((error) => {
    console.error("Échec de l'import du dump", error);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
